Feature/symbolic explanations

sarah-segel opened the pull request:
## Plugins
- Add symbolic explanations plugin (#46).

## Enhancements
- Update author email.
- Add an exit button that first deletes running jobs and then terminates DeepCave (otherwise jobs such as the symbolic regression would continue running after DeepCave termination).
- Nicer handling of Keyboard Interrupt (avoid long error message in terminal).
- Disable debug mode (to avoid error messages being shown directly in the UI, especially upon exit of DeepCave).

## Bug-Fixes
- Fix varying configspace index order in PDP and symbolic explanations.

## Others
- Disabled existing loggers again to suppress redis logs

helegraf commented on Makefile:
```diff
@@ -2,7 +2,7 @@
 
 SHELL := /bin/bash
-VERSION := 1.1.3
+VERSION := 1.2
```
This says version 1.2, but in the Changelog the change is under Version 1.1.4

sarah-segel replied:
Edited to 1.2 in Changelog.

helegraf commented on deepcave/__init__.py:
```diff
@@ -9,15 +9,15 @@
 }
 copyright = f"Copyright {datetime.date.today().strftime('%Y')}, {author}"
-version = "1.1.3"
+version = "1.2"
```
Same as the previous comment about inconsistent versions

sarah-segel replied:
Same as above.

helegraf commented on docs/plugins/symbolic_explanations.rst:
```diff
@@ -0,0 +1,17 @@
+
+This plugin is capable of answering similar questions as the Partial Dependencies plugin, i.e.:
+
+* How does the objective change wrt one or two hyperparameters? For example, does the accuracy
```
I would just write "with respect to"

sarah-segel replied:
Changed that in all cases of wrt

helegraf commented on docs/plugins/symbolic_explanations.rst:
```diff
@@ -0,0 +1,17 @@
+Symbolic Explanations
+====================
+
+This plugin is capable of answering similar questions as the Partial Dependencies plugin, i.e.:
```
I personally think an explanation of what symbolic explanations are in the first sentence would be best; so just adding one more sentence before this one.

sarah-segel replied:
Added a sentence in the beginning.

helegraf commented on deepcave/cli.py:
```diff
@@ -1,5 +1,4 @@
-from typing import Any, List
-
+import os
```
os is unsused

sarah-segel replied:
Removed.

helegraf commented on deepcave/config.py:
```diff
@@ -49,6 +49,9 @@ def PLUGINS(self) -> Dict[str, List["Plugin"]]:
         from deepcave.plugins.hyperparameter.pdp import PartialDependencies
+        from deepcave.plugins.hyperparameter.symbolic_explanations import (
+            SymbolicExplanations,
+        )
```
Why the brackets? :)

sarah-segel replied:
Not sure how they got there, removed :)

helegraf commented on deepcave/utils/symbolic_regression.py:
```diff
@@ -0,0 +1,110 @@
+    try:
+        # Simplification can fail in some cases. If so, use the unsimplified version.
+        symb_simpl = sympy.simplify(symb_conv)
+    except:
```
Is there a specific exception that could be caught here instead?

sarah-segel replied:
I unfortunately didn't track that and cannot find an example now, thus I just changed it to actually print the exception (so if I should see one at some point, I can edit it).

helegraf commented on deepcave/utils/symbolic_regression.py:
```diff
@@ -0,0 +1,110 @@
+
+def convert_symb(symb, n_decimals: int = None, hp_names: list = None) -> sympy.core.expr:
+    """
+    Convert a fitted symbolic regression to a simplified and potentially rounded mathematical expression.
```
This line is over 100 characters; however, I am not 100% sure what our desired setting for the linter would be. 100 is admittedly really short.

helegraf commented on deepcave/utils/symbolic_regression.py:
```diff
@@ -0,0 +1,110 @@
+    return function_set
+
+
+def convert_symb(symb, n_decimals: int = None, hp_names: list = None) -> sympy.core.expr:
```
Type of symb?

sarah-segel replied:
Added.

helegraf commented on deepcave/plugins/hyperparameter/symbolic_explanations.py:
```diff
@@ -0,0 +1,463 @@
+                                [
+                                    dbc.Label("Parsimony coefficient"),
+                                    help_button(
+                                        "Penalizes the complexity of the resulting formulas."
```
Maybe add a line if higher or lower will mean less complexity

sarah-segel replied:
True, added.

helegraf reviewed (commented):
Looks nice overall and the tests are working 💪 I have some minor remarks. It would also be nice if you could run the pre-commit hooks so everything is formatted properly and address the errors in the new files (some minor things like missing docstring I think).

Also is it correct that I am getting DEBUG console output without changing any settings?

Furthermore, I am running into some problems running the plugin but maybe I am using it wrong. When I select the smac-pendigits run, the plugin does not seem to produce any explanations.

Server Log
```
SymbolicExplanations (DEBUG): Button pressed.
SymbolicExplanations (DEBUG): Enqueued SMAC-pendigits-015-0-0 (69666054159540351ed1f6fbf825f2cd).
RandomForestSurrogate (DEBUG): RandomForestSurrogate: Initial Seed = 0
RandomForestSurrogate (DEBUG): RandomForestSurrogate: Final Seed = 248400791
PDP (DEBUG): PDP: Initial Seed = 0
PDP (DEBUG): PDP: Final Seed = 1543538891
ICE (DEBUG): ICE: Initial Seed = True
ICE (DEBUG): ICE: Seed = Use existing
ICE (INFO): Recalculating ICE...
    |   Population Average    |             Best Individual              |
---- ------------------------- ------------------------------------------ ----------
 Gen   Length          Fitness   Length          Fitness      OOB Fitness  Time Left
   0    12.62          181.416       11         0.014005              N/A     17.39s
   1     8.73          20.5379       11         0.014005              N/A     14.85s
   2     7.70          0.39384       10         0.014005              N/A     12.92s
   3     4.63           0.2092       10         0.014005              N/A     11.07s
   4     2.15         0.086137        3        0.0140051              N/A      8.63s
   5     2.02        0.0345978        2        0.0140079              N/A      6.89s
   6     2.04        0.0389469        2        0.0140079              N/A      5.37s
   7     2.03        0.0395785        2        0.0140079              N/A      3.46s
   8     2.05        0.0408348        2        0.0140079              N/A      1.72s
   9     2.03         0.039717        2        0.0140079              N/A      0.00s
SymbolicExplanations (DEBUG): Job 69666054159540351ed1f6fbf825f2cd-48ad708a7a708995d8fdf94b34a2dca3 for run_id 69666054159540351ed1f6fbf825f2cd.
SymbolicExplanations (DEBUG): Job 69666054159540351ed1f6fbf825f2cd-48ad708a7a708995d8fdf94b34a2dca3 cached.
SymbolicExplanations (DEBUG): Job 69666054159540351ed1f6fbf825f2cd-48ad708a7a708995d8fdf94b34a2dca3 deleted.
```

<img width="1840" alt="Bildschirmfoto 2024-02-09 um 11 51 20" src="https://github.com/automl/DeepCAVE/assets/26763001/0b85c51d-e6eb-49fd-9184-79b29a9adf84">

sarah-segel commented:
> Looks nice overall and the tests are working 💪 I have some minor remarks. It would also be nice if you could run the pre-commit hooks so everything is formatted properly and address the errors in the new files (some minor things like missing docstring I think).
> 
> Also is it correct that I am getting DEBUG console output without changing any settings?
> 
> Furthermore, I am running into some problems running the plugin but maybe I am using it wrong. When I select the smac-pendigits run, the plugin does not seem to produce any explanations.
> 
> Server Log
> 
> ```
> SymbolicExplanations (DEBUG): Button pressed.
> SymbolicExplanations (DEBUG): Enqueued SMAC-pendigits-015-0-0 (69666054159540351ed1f6fbf825f2cd).
> RandomForestSurrogate (DEBUG): RandomForestSurrogate: Initial Seed = 0
> RandomForestSurrogate (DEBUG): RandomForestSurrogate: Final Seed = 248400791
> PDP (DEBUG): PDP: Initial Seed = 0
> PDP (DEBUG): PDP: Final Seed = 1543538891
> ICE (DEBUG): ICE: Initial Seed = True
> ICE (DEBUG): ICE: Seed = Use existing
> ICE (INFO): Recalculating ICE...
>     |   Population Average    |             Best Individual              |
> ---- ------------------------- ------------------------------------------ ----------
>  Gen   Length          Fitness   Length          Fitness      OOB Fitness  Time Left
>    0    12.62          181.416       11         0.014005              N/A     17.39s
>    1     8.73          20.5379       11         0.014005              N/A     14.85s
>    2     7.70          0.39384       10         0.014005              N/A     12.92s
>    3     4.63           0.2092       10         0.014005              N/A     11.07s
>    4     2.15         0.086137        3        0.0140051              N/A      8.63s
>    5     2.02        0.0345978        2        0.0140079              N/A      6.89s
>    6     2.04        0.0389469        2        0.0140079              N/A      5.37s
>    7     2.03        0.0395785        2        0.0140079              N/A      3.46s
>    8     2.05        0.0408348        2        0.0140079              N/A      1.72s
>    9     2.03         0.039717        2        0.0140079              N/A      0.00s
> SymbolicExplanations (DEBUG): Job 69666054159540351ed1f6fbf825f2cd-48ad708a7a708995d8fdf94b34a2dca3 for run_id 69666054159540351ed1f6fbf825f2cd.
> SymbolicExplanations (DEBUG): Job 69666054159540351ed1f6fbf825f2cd-48ad708a7a708995d8fdf94b34a2dca3 cached.
> SymbolicExplanations (DEBUG): Job 69666054159540351ed1f6fbf825f2cd-48ad708a7a708995d8fdf94b34a2dca3 deleted.
> ```
> 
> <img alt="Bildschirmfoto 2024-02-09 um 11 51 20" width="1840" src="https://private-user-images.githubusercontent.com/26763001/303628921-0b85c51d-e6eb-49fd-9184-79b29a9adf84.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwNzY2NTUsIm5iZiI6MTcwODA3NjM1NSwicGF0aCI6Ii8yNjc2MzAwMS8zMDM2Mjg5MjEtMGI4NWM1MWQtZTZlYi00OWZkLTkxODQtNzliMjlhOWFkZjg0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE2VDA5MzkxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTliMmFkZmZiMGY1ZTEwYmY5M2YwZGI5YzBjM2JhZTEyNjY4MDY2ZWQxNTEyMjBhZDI2NTE3ZTBiMzAyOWQ2NDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.AyDk1kGfRZEkO83J5BxAkawW3wOG-ozvwS9KmzhK6EU">

Regarding the debug logging, I changed the plugin (and other) specific logging settings in `logging.yml ` to `INFO`.

For the pre-commit hooks, I edited the files I added similar to the changes in the pre-commit branch.

Regarding the problem with the smac-pendigits run, this seems to be related to the problem I ran into with one of Daphne's runs (https://github.com/automl/DeepCAVE/issues/111). I added the example there.

@helegraf  would be great if you could re-check if everything is addressed :)

alexandertornede commented on deepcave/utils/logging.yml:
```diff
@@ -22,6 +22,6 @@ loggers:
+  level: INFO
   handlers: [console]
-disable_existing_loggers: false
+disable_existing_loggers: true
```
What is the reason for us setting this to true?

alexandertornede commented on docs/plugins/configurations.rst:
```diff
@@ -13,7 +13,7 @@ Since configurations are used throughout the application, you might find links a
 
 * Where is the configuration coming from?
-* How are the objective values wrt the budgets?
+* How are the objective values with respect to the budgets?
```
What are the ....

alexandertornede reviewed (approved):
Suggested some small adaptations that might be nice.
